File: pegasusX/packages/mobile-ios-kit/Sources/PegasusKit/Net/ReconnectBackoff.swift
import Foundation

/// Exponential backoff + jitter for WS / telemetry reconnect (§8.8).
public enum ReconnectBackoff {
    public static func delayMs(
        attempt: Int,
        baseMs: Int64 = 2_000,
        maxMs: Int64 = 60_000,
        retryAfterMs: Int64? = nil
    ) -> Int64 {
        let capped = min(max(attempt, 0), 10)
        let exp = min(baseMs * (1 << capped), maxMs)
        let jitter = Int64.random(in: 0...(exp / 2 + 1))
        let jittered = exp + jitter
        return max(jittered, retryAfterMs ?? 0)
    }
}

File: pegasusX/packages/mobile-android-kit/src/main/java/com/pegasusx/mobilekit/net/ReconnectBackoff.kt
package com.pegasusx.mobilekit.net

import kotlin.math.min
import kotlin.random.Random

/** Exponential backoff + jitter for WS / telemetry reconnect (§8.8). */
object ReconnectBackoff {
    fun delayMs(
        attempt: Int,
        baseMs: Long = 2_000L,
        maxMs: Long = 60_000L,
        retryAfterMs: Long? = null,
    ): Long {
        val capped = attempt.coerceIn(0, 10)
        val exp = min(baseMs * (1L shl capped), maxMs)
        val jittered = exp + Random.nextLong(0, exp / 2 + 1)
        return maxOf(jittered, retryAfterMs ?: 0L)
    }

    fun retryAfterHeaderSeconds(raw: String?): Long? =
        raw?.toLongOrNull()?.times(1_000)
}
